### ISSUE TITLE
Vagrant workder error are not reported back to the user

### ISSUE THREAD

**Issue by pbuyle** [StudioQi/jeto#57]
In addition to not letting the user known something went wrong (see #52), error in vagrant worker are not forwarded to the client, so the user has not way of knowing if she did something wrong or if there is a bug in Jeto, which make usage of Jeto quite uncomfortable.

**Comment by abondis** [comment]
maybe this can help http://docs.vagrantup.com/v2/cli/machine-readable.html

**Comment by abondis** [comment]
see  #54 and #49 too

**Comment by pbuyle** [comment]
Quick review of the code reveal that many calls to API are still no checked for return conditions.

https://github.com/StudioQi/jeto/blob/master/jeto/static/js/controllers/instance.js#L51
=> No check for error when retrieving a machine IP

https://github.com/StudioQi/jeto/blob/master/jeto/static/js/controllers/instance.js#L98
=> No check for error when deleting an instance

https://github.com/StudioQi/jeto/blob/master/jeto/static/js/controllers/instance.js#L130
=> No check for error when getting a job's details

Really guys, we need to have some kind of code review and get into the practice of checking for error. In a networked complex application, errors happen and it should be the norm to check for them, not the exception.

**Comment by PierrePaul** [comment]
Hi @pbuyle ! Thank you for finding those ill-mannered calls. Code is always reviewed every time a pull request is made and specific issues are tested against the pull request. If the issue is too vague (like this one), a specific test will be made and a full test done to make sure everything still works fine.

If you want to give a hand, I invite you to read the github guide about contributing (https://guides.github.com/activities/contributing-to-open-source/). It highlights how to create a local clone, make changes and create a pull request.

I'm reopening this issue as you found some unhandled cases. Keep in mind that people working on this project are doing this in their free time and doing the best they can to deliver a quality product.

**Comment by pbuyle** [comment]
This issue is not vague, it is just not a bug report. It is not a request for a specific problem to be fixed, it point to specific anity-patterns replicated all over the application. I don't remember being unable to find code to forward low level back-end error (eg. errors when invoking Vagrant commands) to the end user, but it may be that Flask exposes Python exceptions directly as HTTP error (code 5xx with a descriptive body?). But in itself exposing back-end errors to the front-end is useless, since most client-side code does not check for error.

AFAIK, these two anti-patterns have not been acknowledge for. Fixing individual occurrence one by one would be nice. But having a clear documentation/knowledge that they should be avoided in the first place and how to properly forward back-end errors to the client, and how to report those errors to the end user is needed first. Otherwise, you are just fixings symptoms, not curing the illness.

For instance, in commit https://github.com/StudioQi/jeto/commit/24464d8c20d30eac05c457282967adebb7c03cd6, several API request have been added in JavaScript code, without any error handling code.

I'm not sure how to translate this into actionable items. However, reporting individual missing error handling on the front-end to get them fixed individually is a painful waste of time without a clear vision on how error are to be handled in the front-end. Do we want to replicate the error handling found at https://github.com/StudioQi/jeto/blob/master/jeto/static/js/controllers/instance.js#L38 (an alert box with a non-standardized error message)? Can a general error default handler be registered for all model methods and HTTP requests?

**Comment by PierrePaul** [comment]
> Do we want to replicate the error handling found at https://github.com/StudioQi/jeto/blob/master/jeto/static/js/controllers/instance.js#L38 (an alert box with a non-standardized error message)?

Now that's an excellent actionable item. Have a generic way to handle errors and still keep a precise error message. And of course, document it and use it.

**Comment by PierrePaul** [comment]
Linking to #74 for documenting purposes.

**Comment by pbuyle** [comment]
It should be possible to have a *default* error handler for HTTP request, see http://stackoverflow.com/questions/17257946/angular-default-handler-for-unhandled-http-errors

Because resource objects are also using HTTP, it should intercept HTTP for them too.

**Comment by PierrePaul** [comment]
Exactly, but hacking $http seems like a bad long term idea. I will start working on an $interceptor tonight.
